File: java/diamond/src/main/java/DiamondPrinter.java
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

class DiamondPrinter {

    private static final char BASE = 'A';

    List<String> printToList(char a) {
        int size = a - BASE + 1;

        // create the top right quadrant
        char[][] quadrant = new char[size][size];
        for (int i = 0; i < size; i++) {
            Arrays.fill(quadrant[i], ' ');
            quadrant[i][i] = (char)(BASE + i);
        }

        // create the top half
        List<String> diamond = Arrays
                .stream(quadrant)
                .map(String::new)
                .map(s -> new StringBuilder(s).deleteCharAt(0).reverse().append(s))
                .map(StringBuilder::toString)
                .collect(Collectors.toList());

        // add the bottom half
        for (int i = size - 2; i >= 0; i--) {
            diamond.add(diamond.get(i));
        }

        return diamond;
    }

}

File: kotlin/diamond/src/main/kotlin/DiamondPrinter.kt
class DiamondPrinter {

    private fun row(n: Int, i: Int): String =
        with( " ".repeat(i) + ('A' + i).toString() + " ".repeat(n - i - 1)) {
            this.drop(1).reversed() + this
        }

    fun printToList(c: Char): List<String> =
        (c - 'A' + 1).let { n ->
            require(c in 'A'..'Z')
            with(
                (0 until n).map { row(n, it) }
            ) {
                this + this.reversed().drop(1)
            }
        }
}
